src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiScriptTransformer.cs

```diff
@@ -25,38 +28,98 @@ internal sealed class GenAiScriptTransformer : EtlTransformer<GenAiItem, GenAiSc
+    private static readonly string JavaScriptApi = @"
+class AIContextItem {
+  #withAttachment(type, data) {
+    if (typeof type !== 'string' || (typeof data !== 'string' && data !== null)) {
+        throw new Error('both type and data must be strings');
+    }
+    this.attachments.push({ type, data });
+    return this;
+  }
+
+  constructor(ctx) {
+    if (ctx !== null && (typeof ctx !== 'object' || Array.isArray(ctx))) {
+      throw new Error('ctx must be an object');
+    }
+    this.ctx = ctx;
+    this.attachments = [];
+  }
+
+  withText(data) {
+    return this.#withAttachment('text/plain', data);
+  }
+
+  withJpeg(data) {
+    return this.#withAttachment('image/jpeg', data);
+  }
+
+  withPng(data) {
+    return this.#withAttachment('image/png', data);
+  }
+
+  withWebp(data) {
+    return this.#withAttachment('image/webp', data);
+  }
+
+  withGif(data) {
+    return this.#withAttachment('image/gif', data);
+  }
+
+  withPdf(data) {
+    return this.#withAttachment('application/pdf', data);
+  }
+}
+
+class AI {
+  #allContexts = [];
+
+  __retrieveContexts() {
+     const ctxs = this.#allContexts;
+     this.#allContexts = [];
+     return ctxs;
+  }
+
+  genContext(...args) {
+    if (args.length !== 1) {
+      throw new Error('Invalid number of arguments for ai.genContext(ctx)');
+    }
+    const ctx = new AIContextItem(args[0]);
+    this.#allContexts.push(ctx);
+    return ctx;
+  }
+}
+
+var ai = new AI();
+"; 
+
+    public GenAiScriptTransformer(DocumentDatabase database, DocumentsOperationContext context, Transformation transformation, PatchRequest behaviorFunctions, GenAiConfiguration configuration, GenAiStatsScope stats) 
+        : base(database, context, new PatchRequest(transformation.Script, PatchRequestType.GenAi), behaviorFunctions)
     {
         _configuration = configuration;
         _stats = stats.For(EtlOperations.Transform, start: false);
-        _mainScript = new PatchRequest(transformation.Script, PatchRequestType.GenAi);
     }
 
     public override void Initialize(bool debugMode)
     {
-        ReturnMainRun = Database.Scripts.GetScriptRunner(_mainScript, true, out DocumentScript);
-
-        if (DocumentScript == null)
-            return;
-
-        if (debugMode)
-            DocumentScript.DebugMode = true;
-
-        var contextFunc = new ClrFunction(DocumentScript.ScriptEngine, "genContext", AddContext);
-        ObjectInstance aiObject = new JsObject(DocumentScript.ScriptEngine);
-        aiObject.FastSetProperty("genContext", new PropertyDescriptor(contextFunc, false, false, false));
-        DocumentScript.ScriptEngine.SetValue("ai", aiObject);
-
         _configurationPartialHash = GetInitialHash(_configuration);
+
+        base.Initialize(debugMode);
+        JsValue aiAlreadyExists = DocumentScript.ScriptEngine.GetValue("ai");
+        if (aiAlreadyExists.IsNull() || aiAlreadyExists.IsUndefined())
+        {
+            DocumentScript.ScriptEngine.Execute(JavaScriptApi);
```

Comment:
FYI, this is probably slower, not a concern for GenAI, but now we are executing Jint twice